File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/settings/ManualAddSearchEngineSettingsFragment.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.settings;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import org.mozilla.focus.R;
import org.mozilla.focus.activity.InfoActivity;
import org.mozilla.focus.search.ManualAddSearchEnginePreference;
import org.mozilla.focus.search.SearchEngineManager;
import org.mozilla.focus.telemetry.TelemetryWrapper;
import org.mozilla.focus.utils.SupportUtils;
import org.mozilla.focus.utils.UrlUtils;
import org.mozilla.focus.utils.ViewUtils;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

public class ManualAddSearchEngineSettingsFragment extends SettingsFragment {
    private static String LOGTAG = "ManualAddSearchEngine";

    // Set so the user doesn't have to wait *too* long. It's used twice: once for connecting and once for reading.
    private static int SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS = 4000;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onResume() {
        super.onResume();

        // We've checked that this cast is legal in super.onAttach.
        ((ActionBarUpdater) getActivity()).updateIcon(R.drawable.ic_close);
   }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_search_engine_manual_add, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.learn_more:
                final Context context = getActivity();
                final String url = SupportUtils.getSumoURLForTopic(context, "add-search-engine");

                final String title = ((AppCompatActivity) getActivity()).getSupportActionBar().getTitle().toString();
                final Intent intent = InfoActivity.getIntentFor(context, url, title);
                context.startActivity(intent);

                TelemetryWrapper.addSearchEngineLearnMoreEvent();
                return true;

            case R.id.menu_save_search_engine:
                final View rootView = getView();
                final String engineName = ((EditText) rootView.findViewById(R.id.edit_engine_name)).getText().toString();
                final String searchQuery = ((EditText) rootView.findViewById(R.id.edit_search_string)).getText().toString();

                final ManualAddSearchEnginePreference pref = (ManualAddSearchEnginePreference) findPreference(getString(R.string.pref_key_manual_add_search_engine));
                final boolean engineValid = pref.validateEngineNameAndShowError(engineName);
                final boolean searchValid = pref.validateSearchQueryAndShowError(searchQuery);
                final boolean isPartialSuccess = engineValid && searchValid;

                if (isPartialSuccess) {
                    // Hide the keyboard because:
                    // - It's awkward to show the keyboard while waiting for a response
                    // - We want it hidden when we return to the previous screen (on success)
                    // - An expanded keyboard hides the success snackbar
                    ViewUtils.hideKeyboard(rootView);
                    setUiIsValidatingAsync(true, item);
                    new ValidateSearchEngineAsyncTask(this, item, engineName, searchQuery).execute();
                } else {
                    TelemetryWrapper.saveCustomSearchEngineEvent(false);
                }
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void setUiIsValidatingAsync(final boolean isValidating, final MenuItem saveMenuItem) {
        final ManualAddSearchEnginePreference pref =
                (ManualAddSearchEnginePreference) findPreference(getString(R.string.pref_key_manual_add_search_engine));
        pref.setProgressViewShown(isValidating);

        saveMenuItem.setEnabled(!isValidating);
    }

    private static class ValidateSearchEngineAsyncTask extends AsyncTask<Void, Void, Boolean> {
        private final WeakReference<ManualAddSearchEngineSettingsFragment> thatWeakReference;
        private final WeakReference<MenuItem> saveMenuItemWeakReference;
        private final String engineName;
        private final String query;

        private ValidateSearchEngineAsyncTask(final ManualAddSearchEngineSettingsFragment that, final MenuItem saveMenuItem,
                final String engineName, final String query) {
            this.thatWeakReference = new WeakReference<>(that);
            this.saveMenuItemWeakReference = new WeakReference<>(saveMenuItem); // contains reference to Context.
            this.engineName = engineName;
            this.query = query;
        }

        @Override
        protected Boolean doInBackground(final Void... voids) {
            final boolean isValidSearchQuery = isValidSearchQueryURL(query);
            TelemetryWrapper.saveCustomSearchEngineEvent(isValidSearchQuery);
            return isValidSearchQuery;
        }

        @Override
        protected void onPostExecute(final Boolean isValidSearchQuery) {
            super.onPostExecute(isValidSearchQuery);

            final ManualAddSearchEngineSettingsFragment that = thatWeakReference.get();
            final MenuItem saveMenuItem = saveMenuItemWeakReference.get();
            if (that == null || saveMenuItem == null) {
                Log.d(LOGTAG, "Fragment or menu item no longer exists when search query validation async task returned.");
                return;
            }

            if (isValidSearchQuery) {
                final SharedPreferences sharedPreferences = that.getSearchEngineSharedPreferences();
                SearchEngineManager.addSearchEngine(sharedPreferences, that.getActivity(), engineName, query);
                Snackbar.make(that.getView(), R.string.search_add_confirmation, Snackbar.LENGTH_SHORT).show();
                that.getFragmentManager().popBackStack();
            } else {
                showServerError(that);
            }

            that.setUiIsValidatingAsync(false, saveMenuItem);
        }

        private void showServerError(final ManualAddSearchEngineSettingsFragment that) {
            final ManualAddSearchEnginePreference pref = (ManualAddSearchEnginePreference) that.findPreference(
                    that.getString(R.string.pref_key_manual_add_search_engine));
            pref.setSearchQueryErrorText(that.getString(R.string.error_hostLookup_title));
        }
    }

    @WorkerThread // makes network request.
    private static boolean isValidSearchQueryURL(final String query) {
        // TODO: we should share the code to substitute and normalize the search string (see SearchEngine.buildSearchUrl).
        final String encodedTestQuery = Uri.encode("test");
        final String normalizedSearchURLStr = UrlUtils.normalize(query);
        final String searchURLStr = normalizedSearchURLStr.replaceAll("%s", encodedTestQuery);

        final URL searchURL;
        try {
            searchURL = new URL(searchURLStr);
        } catch (final MalformedURLException e) {
            // Don't log exception to avoid leaking URL.
            Log.d(LOGTAG, "Malformed URL: returning invalid search query");
            return false;
        }

        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) searchURL.openConnection();
            connection.setConnectTimeout(SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS);
            connection.setReadTimeout(SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS);

            // Checking for 200 is not perfect - what if a search engine redirects us? -
            // but we're in a rush and it's good enough for now. One notable case this fails:
            // if sites redirect to https, e.g. "duckduckgo.com/?q=%s" redirects to https and thus fails.
            return connection.getResponseCode() == 200;

        } catch (final IOException e) {
            // Don't log exception to avoid leaking URL.
            Log.d(LOGTAG, "Failure to get response code from server: returning invalid search query");
            return false;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }
}

File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/settings/ManualAddSearchEngineSettingsFragment.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.settings

import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.annotation.WorkerThread
import androidx.core.view.forEach
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import mozilla.components.browser.icons.IconRequest
import mozilla.components.browser.state.state.searchEngines
import mozilla.components.concept.fetch.Client
import mozilla.components.concept.fetch.Request
import mozilla.components.concept.fetch.Request.Redirect.FOLLOW
import mozilla.components.feature.search.ext.createSearchEngine
import mozilla.components.service.glean.private.NoExtras
import mozilla.components.support.ktx.util.URLStringUtils
import org.mozilla.focus.GleanMetrics.SearchEngines
import org.mozilla.focus.R
import org.mozilla.focus.ext.components
import org.mozilla.focus.ext.requireComponents
import org.mozilla.focus.ext.settings
import org.mozilla.focus.ext.showToolbar
import org.mozilla.focus.search.ManualAddSearchEnginePreference
import org.mozilla.focus.state.AppAction
import org.mozilla.focus.utils.SupportUtils
import org.mozilla.focus.utils.ViewUtils
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL
import java.util.concurrent.TimeUnit

@Suppress("TooManyFunctions")
class ManualAddSearchEngineSettingsFragment : BaseSettingsFragment() {
    override fun onCreatePreferences(p0: Bundle?, p1: String?) {
        addPreferencesFromResource(R.xml.manual_add_search_engine)
    }

    private var scope: CoroutineScope? = null
    private var menuItemForActiveAsyncTask: MenuItem? = null
    private var job: Job? = null

    override fun onResume() {
        super.onResume()

        showToolbar(getString(R.string.action_option_add_search_engine))
    }

    override fun onPause() {
        super.onPause()
        setUiIsValidatingAsync(false, menuItemForActiveAsyncTask)
        menuItemForActiveAsyncTask = null
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.menu_search_engine_manual_add, menu)
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        val openLearnMore = {
            val learnMoreUrl = SupportUtils.getSumoURLForTopic(
                SupportUtils.getAppVersion(requireContext()),
                SupportUtils.SumoTopic.ADD_SEARCH_ENGINE,
            )
            SupportUtils.openUrlInCustomTab(requireActivity(), learnMoreUrl)
            SearchEngines.learnMoreTapped.record(NoExtras())

            true
        }

        val saveSearchEngine = {
            val engineName = requireView().findViewById<EditText>(R.id.edit_engine_name).text.toString()
            val searchQuery = requireView().findViewById<EditText>(R.id.edit_search_string).text.toString()

            val pref = findManualAddSearchEnginePreference(R.string.pref_key_manual_add_search_engine)

            val existingEngines = requireContext().components.store.state.search.searchEngines
            val engineValid = pref?.validateEngineNameAndShowError(engineName, existingEngines) ?: false
            val searchValid = pref?.validateSearchQueryAndShowError(searchQuery) ?: false
            val isPartialSuccess = engineValid && searchValid

            if (isPartialSuccess) {
                ViewUtils.hideKeyboard(view)
                setUiIsValidatingAsync(true, menuItem)

                menuItemForActiveAsyncTask = menuItem
                scope?.launch {
                    validateSearchEngine(engineName, searchQuery, requireComponents.client)
                }
            } else {
                SearchEngines.saveEngineTapped.record(SearchEngines.SaveEngineTappedExtra(false))
            }

            true
        }

        return when (menuItem.itemId) {
            R.id.learn_more -> openLearnMore()
            R.id.menu_save_search_engine -> saveSearchEngine()
            else -> false
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        scope = CoroutineScope(Dispatchers.IO)
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onDestroyView() {
        scope?.cancel()
        super.onDestroyView()
        if (view != null) ViewUtils.hideKeyboard(view)
    }

    private fun setUiIsValidatingAsync(isValidating: Boolean, saveMenuItem: MenuItem?) {
        val pref = findManualAddSearchEnginePreference(R.string.pref_key_manual_add_search_engine)
        val updateViews = {
            // Disable text entry until done validating
            val viewGroup = view as ViewGroup
            enableAllSubviews(!isValidating, viewGroup)

            saveMenuItem?.isEnabled = !isValidating
        }

        if (isValidating) {
            view?.alpha = DISABLED_ALPHA
            // Delay showing the loading indicator to prevent it flashing on the screen
            job = scope?.launch(Dispatchers.Main) {
                delay(LOADING_INDICATOR_DELAY)
                pref?.setProgressViewShown(isValidating)
                updateViews()
            }
        } else {
            view?.alpha = 1f
            job?.cancel()
            pref?.setProgressViewShown(isValidating)
            updateViews()
        }
    }

    private fun enableAllSubviews(shouldEnable: Boolean, viewGroup: ViewGroup) {
        viewGroup.forEach { child ->
            if (child is ViewGroup) {
                enableAllSubviews(shouldEnable, child)
            } else {
                child.isEnabled = shouldEnable
            }
        }
    }

    private fun findManualAddSearchEnginePreference(id: Int): ManualAddSearchEnginePreference? {
        return findPreference(getString(id)) as? ManualAddSearchEnginePreference
    }

    companion object {
        private const val LOGTAG = "ManualAddSearchEngine"
        private const val SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS = 4000
        private const val VALID_RESPONSE_CODE_UPPER_BOUND = 300
        private const val DISABLED_ALPHA = 0.5f
        private const val LOADING_INDICATOR_DELAY: Long = 1000

        @WorkerThread
        @JvmStatic
        fun isValidSearchQueryURL(client: Client, query: String): Boolean {
            // we should share the code to substitute and normalize the search string (see SearchEngine.buildSearchUrl).
            val encodedTestQuery = Uri.encode("testSearchEngineValidation")

            val normalizedHttpsSearchURLStr = URLStringUtils.toNormalizedURL(query)
            val searchURLStr = normalizedHttpsSearchURLStr.replace("%s".toRegex(), encodedTestQuery)

            try { URL(searchURLStr) } catch (e: MalformedURLException) {
                // Don't log exception to avoid leaking URL.
                Log.d(LOGTAG, "Failure to get response code from server: returning invalid search query")
                return false
            }

            val request = Request(
                url = searchURLStr,
                connectTimeout = SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS.toLong() to TimeUnit.MILLISECONDS,
                readTimeout = SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS.toLong() to TimeUnit.MILLISECONDS,
                redirect = FOLLOW,
                private = true,
            )

            return try {
                val response = client.fetch(request)
                // Close the response stream to ensure the body is closed correctly. See https://bugzilla.mozilla.org/show_bug.cgi?id=1603114.
                response.close()

                response.status < VALID_RESPONSE_CODE_UPPER_BOUND
            } catch (e: IOException) {
                Log.d(LOGTAG, "Failure to get response code from server: returning invalid search query")
                false
            }
        }
    }

    private suspend fun validateSearchEngine(engineName: String, query: String, client: Client) {
        val isValidSearchQuery = isValidSearchQueryURL(client, query)

        withContext(Dispatchers.Main) {
            if (!isActive) {
                return@withContext
            }

            if (isValidSearchQuery) {
                requireComponents.searchUseCases.addSearchEngine(
                    createSearchEngine(
                        engineName,
                        query.toSearchUrl(),
                        requireComponents.icons.loadIcon(IconRequest(query, isPrivate = true)).await().bitmap,
                    ),
                )

                ViewUtils.showBrandedSnackbar(requireView(), R.string.search_add_confirmation, Snackbar.LENGTH_SHORT)
                requireActivity().settings.setDefaultSearchEngineByName(engineName)
                SearchEngines.saveEngineTapped.record(SearchEngines.SaveEngineTappedExtra(true))

                requireComponents.appStore.dispatch(
                    AppAction.NavigateUp(requireComponents.store.state.selectedTabId),
                )
            } else {
                showServerError()
                SearchEngines.saveEngineTapped.record(SearchEngines.SaveEngineTappedExtra(false))
            }

            setUiIsValidatingAsync(false, menuItemForActiveAsyncTask)
            menuItemForActiveAsyncTask = null
        }
    }

    private fun showServerError() {
        val pref = findManualAddSearchEnginePreference(R.string.pref_key_manual_add_search_engine)
        pref?.setSearchQueryErrorText(getString(R.string.error_hostLookup_title))
    }
}

private fun String.toSearchUrl(): String {
    return replace("%s", "{searchTerms}")
}
